File: packages/sdk-ruby/lib/pocketping/ai/anthropic_provider.rb
# frozen_string_literal: true

require_relative "base"

module PocketPing
  module AI
    # Anthropic Claude Messages provider.
    #
    # @example
    #   provider = PocketPing::AI::AnthropicProvider.new(api_key: ENV["ANTHROPIC_API_KEY"])
    #   client = PocketPing::Client.new(ai_provider: provider)
    class AnthropicProvider < Base
      DEFAULT_MODEL = "claude-sonnet-4-20250514"
      DEFAULT_BASE_URL = "https://api.anthropic.com/v1"
      DEFAULT_SYSTEM_PROMPT = "You are a helpful customer support assistant."
      ANTHROPIC_VERSION = "2023-06-01"

      # @param api_key [String] Anthropic API key (required)
      # @param model [String] Model name (default: "claude-sonnet-4-20250514")
      # @param base_url [String] API base URL (default: "https://api.anthropic.com/v1")
      def initialize(api_key:, model: DEFAULT_MODEL, base_url: DEFAULT_BASE_URL)
        super()
        raise ArgumentError, "api_key is required" if api_key.nil? || api_key.empty?

        @api_key = api_key
        @model = model
        @base_url = base_url.chomp("/")
      end

      # @return [String] "anthropic"
      def name
        "anthropic"
      end

      # @param messages [Array<PocketPing::Message>]
      # @param system_prompt [String, nil]
      # @return [String]
      def generate_response(messages, system_prompt = nil)
        chat_messages = messages.map do |message|
          { role: role_for(message), content: message.content }
        end

        body = {
          model: @model,
          max_tokens: 1000,
          system: system_prompt || DEFAULT_SYSTEM_PROMPT,
          messages: chat_messages
        }

        data = post_json("#{@base_url}/messages", body, headers)
        data.dig("content", 0, "text") || ""
      end

      # Anthropic has no health endpoint; available when an API key is set.
      #
      # @return [Boolean]
      def available?
        !@api_key.nil? && !@api_key.empty?
      end

      private

      def headers
        {
          "x-api-key" => @api_key,
          "anthropic-version" => ANTHROPIC_VERSION
        }
      end
    end
  end
end

File: packages/sdk-php/src/AI/AnthropicProvider.php
<?php

declare(strict_types=1);

namespace PocketPing\AI;

use PocketPing\Http\CurlHttpClient;
use PocketPing\Http\HttpClientInterface;
use PocketPing\Models\Message;
use PocketPing\Models\Sender;

/**
 * AI provider backed by the Anthropic Messages API.
 */
final class AnthropicProvider implements AIProviderInterface
{
    public const DEFAULT_BASE_URL = 'https://api.anthropic.com/v1';
    public const DEFAULT_MODEL = 'claude-sonnet-4-20250514';
    public const DEFAULT_SYSTEM_PROMPT = 'You are a helpful customer support assistant.';

    private readonly string $apiKey;
    private readonly string $model;
    private readonly string $baseUrl;
    private readonly HttpClientInterface $httpClient;

    public function __construct(
        string $apiKey,
        string $model = self::DEFAULT_MODEL,
        string $baseUrl = self::DEFAULT_BASE_URL,
        ?HttpClientInterface $httpClient = null,
    ) {
        $this->apiKey = $apiKey;
        $this->model = $model;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->httpClient = $httpClient ?? new CurlHttpClient();
    }

    public function name(): string
    {
        return 'anthropic';
    }

    public function generateResponse(array $messages, ?string $systemPrompt = null): string
    {
        // System prompt is sent in the top-level `system` field, never in the array.
        $anthropicMessages = [];
        foreach ($messages as $msg) {
            $role = $msg->sender === Sender::VISITOR ? 'user' : 'assistant';
            $anthropicMessages[] = ['role' => $role, 'content' => $msg->content];
        }

        $response = $this->httpClient->post(
            "{$this->baseUrl}/messages",
            [
                'model' => $this->model,
                'max_tokens' => 1000,
                'system' => $systemPrompt ?? self::DEFAULT_SYSTEM_PROMPT,
                'messages' => $anthropicMessages,
            ],
            [
                'x-api-key' => $this->apiKey,
                'anthropic-version' => '2023-06-01',
            ],
        );

        if ($response['error'] !== null || $response['httpCode'] < 200 || $response['httpCode'] >= 300) {
            throw new \RuntimeException(
                sprintf('Anthropic API error: %s', $response['error'] ?? "HTTP {$response['httpCode']}")
            );
        }

        $data = json_decode((string) $response['body'], true);
        if (!is_array($data)) {
            return '';
        }

        return (string) ($data['content'][0]['text'] ?? '');
    }

    public function isAvailable(): bool
    {
        // Anthropic has no simple health endpoint; assume available if an API key is set.
        return $this->apiKey !== '';
    }
}
